### DISCUSSION TITLE
Collections divided into groups

### DISCUSSION THREAD

**Discussion by hanqiushi** [milvus-io/milvus#20235] [Q&A and General discussion] [answered]
We have a function like this: Search vector in specific group, e.g. we have a param group, all groups [1,2,3,4], item A is in [1,2], item B in [2,3], when we search in group 3, result may contains B, but not including A.

For milvus, what is the best way to implement this function? Now I don't have a better solution, I can only add  vector A twice (group=1, group = 2), item B twice (2,3), which costs more memory & storage.

**Comment by yhmo** [comment, marked as answer]
Let's say we have several groups of vectors:
group_1: [v1, v2, v3, ... v1000]
group_2: [v500, v501, .... v1500]
group_3: [v1000, v1001, ... v2000]
Each group has 1000 vectors. since there are intersections, actually there are only 2000 unique vectors.

It depends on how your business will query the groups.
If query the groups one by one, then you can create 3 collections, one collection for one group.

If you want to put the vectors into a single collection to search, then we can use an integer field to mark the group. We can create a collection with three fields: "ID", "Vector", "GroupFlag", then we insert the 2000 unique vectors into the collection with their group flag. For example, the v1 is belong to group_1, then its GroupFlag is 1; the v1000 belong to group_1&group_2&group_3, then its GroupFlag is 1|2|4=7; the v500 is belong to group_1&group_2, then its GroupFlag is 1|2=3, etc.
Now you want to search the group_1 and group_2, you can set the parameter 'expr' to be "GroupFlag==3":
```
search_param = {
        "data": search_vectors,
        "anns_field": vector_field,
        "param": {"metric_type": "L2", "params": {"nprobe": 10}},
        "limit": 10,
        "expr": "GroupFlag==3"}
    results = collection.search(**search_param)
```
With this expression, the search() will only search within group_1 and group_2. This is a kind of search we called "hybrid search", it save your CPU memory and storage, but the performance of hybrid search is worse than pure vector search
